### PR TITLE
Properly print cstr literals in `proc_macro::Literal::to_string`

### LINKED ISSUE #112820: `c"…"` literals are handled incorrectly by proc_macro::Literal
```toml
# Cargo.toml

[package]
name = "repro"
version = "0.0.0"
edition = "2021"

[lib]
proc-macro = true
```

```rust
// src/main.rs

#![feature(c_str_literals)]

repro::repro!(b"bytestr" c"cstr");

fn main() {}
```

```rust
// src/lib.rs

use proc_macro::{TokenStream, TokenTree};

#[proc_macro]
pub fn repro(input: TokenStream) -> TokenStream {
    println!("{:#?}", input);
    for token in input {
        if let TokenTree::Literal(literal) = token {
            println!("{}", literal);
        }
    }
    TokenStream::new()
}
```

Current output as of nightly-2023-06-20:

```console
$ cargo check
TokenStream [
    Literal {
        kind: ByteStr,
        symbol: "bytestr",
        suffix: None,
        span: #0 bytes(43..53),
    },
    Literal {
        kind: CStr,
        symbol: "cstr",
        suffix: None,
        span: #0 bytes(54..61),
    },
]
b"bytestr"
cstr
```

The correct output would have `c"cstr"`.

The ToString of proc_macro::Literal is the only interface by which proc macros are able to inspect literals. It is supposed to include the entire literal, with quotes and escape sequences and everything as originally appear in the source code. Without this, it is impossible for proc macros to operate correctly on `c"…"` literals. For example, it would currently be impossible for a proc macro to tell the difference between `b'B'` and `c"b'B'"`, but those are not the same token.

Mentioning tracking issue: https://github.com/rust-lang/rust/issues/105723.
Mentioning @fee1-dead since you have worked on this feature in the past.

### PR DESCRIPTION
**Pull request by WaffleLapkin**
Previously we printed the contents of the string, rather than the actual string literal (e.g. `the c string` instead of `c"the c string"`).

Fixes #112820
cc #105723

### CODE REVIEW

**Comment by compiler-errors** [comment]
r? @compiler-errors @bors r+

Thanks @WaffleLapkin

**Comment by compiler-errors** [comment]
cc @rust-lang/wg-macros since y'all mentioned wanting to be in the loop for libprocmacro-related changes, at least until there is auto-ping group for the WG.

File: library/proc_macro/src/lib.rs [hunk 1]
```diff
@@ -1411,7 +1411,15 @@ impl Literal {
                 let hashes = get_hashes_str(n);
                 f(&["br", hashes, "\"", symbol, "\"", hashes, suffix])
             }
-            _ => f(&[symbol, suffix]),
```
Comment (compiler-errors) [inline library/proc_macro/src/lib.rs:1414]: making macros exhaustively match ❤️

**Review by vincenzopalazzo — COMMENTED** [review]
Do you have any idea why the CI is failing with `thread 'main' has overflowed its stack`?

File: tests/ui/proc-macro/literal-to-string.stdout [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+b'\xFF'
+'c'
+'\x32'
+"\"str\""
+r#""raw" str"#
+r###"very ##"raw"## str"###
+b"\"byte\" str"
+br#""raw" "byte" str"#
+c"\"c\" str"
+cr#""raw" "c" str"#
```
Comment (vincenzopalazzo) [inline tests/ui/proc-macro/literal-to-string.stdout:107]: Running the test on my system looks like that I the following diff

```diff
diff --git a/tests/ui/proc-macro/literal-to-string.stdout b/tests/ui/proc-macro/literal-to-string.stdout
index 7b27fcf798b..78ae480a6b7 100644
--- a/tests/ui/proc-macro/literal-to-string.stdout
+++ b/tests/ui/proc-macro/literal-to-string.stdout
@@ -103,5 +103,5 @@ r#""raw" str"#
 r###"very ##"raw"## str"###
 b"\"byte\" str"
 br#""raw" "byte" str"#
-c"\"c\" str"
-cr#""raw" "c" str"#
+\"c\" str
+"raw" "c" str

```

Comment (compiler-errors) [reply tests/ui/proc-macro/literal-to-string.stdout:107]: Did you apply the changes from PR correctly? This is what the output should look like *before* the changes in this PR.

Comment (vincenzopalazzo) [reply tests/ui/proc-macro/literal-to-string.stdout:107]: You are right was a not clean build

**Comment by WaffleLapkin** [comment]
@compiler-errors does this need a rebase after #116147 or just @bors retry should be enough?

**Comment by compiler-errors** [comment]
We're good

@bors retry

**Pull request merged**